packages/text-field/src/TextField/TextField.tsx

```diff
@@ -0,0 +1,135 @@
+import { useComposedRefs } from "@telegraph/compose-refs";
+import { Icon } from "@telegraph/icon";
+import { clsx } from "clsx";
+import React from "react";
+
+import { COLOR, SIZE } from "./TextField.constants";
+
+type TextFieldProps = Omit<
+  React.HTMLAttributes<HTMLInputElement>,
+  "value" | "onChange"
+> & {
+  size?: "1" | "2" | "3";
+  variant?: "outline" | "ghost";
+  disabled?: boolean;
+  error?: boolean | string;
+  value?: string;
+  onChange?: (
+    value: string,
+    { event }: { event: React.ChangeEvent<HTMLInputElement> },
+  ) => void;
+  leadingIcon?: React.ComponentProps<typeof Icon>;
+  TrailingAction?: React.ReactElement;
+};
+
+type TextFieldRef = HTMLInputElement;
+
+type InternalProps = {
+  id: string;
+  state: "default" | "error" | "disabled";
+  size: TextFieldProps["size"];
+  variant: TextFieldProps["variant"];
+  message?: string;
+};
+
+const TextField = React.forwardRef<TextFieldRef, TextFieldProps>(
+  (
+    {
+      size = "2",
+      variant = "outline",
+      value,
+      onChange,
+      leadingIcon,
+      TrailingAction,
+      className,
+      disabled,
+      error,
+      ...props
+    },
+    forwardedRef,
+  ) => {
+    const inputRef = React.useRef<HTMLInputElement>(null);
+    const composedRefs = useComposedRefs(inputRef, forwardedRef);
+
+    const [internalState, setInternalState] =
+      React.useState<InternalProps["state"]>("default");
+
+    React.useEffect(() => {
+      if (disabled) {
+        setInternalState("disabled");
+      } else if (error) {
+        setInternalState("error");
+      } else {
+        setInternalState("default");
+      }
+    }, [disabled, error]);
+
+    // Adds default props to the component, which can be overridden by the caller
+    const TrailingActionComponent = TrailingAction
+      ? React.cloneElement(TrailingAction, {
+          size,
+          color: "gray",
+          variant: "ghost",
+          ...TrailingAction?.props,
+        })
+      : null;
+    return (
+      <div
+        className={clsx(
+          "box-border flex items-center transition-all",
+          "border-2 border-solid text-gray-12 placeholder:text-gray-10",
+          COLOR.Container[internalState][variant],
+          SIZE.Container[size],
+        )}
+        // Focus the input when clicking on the container
+        onPointerDown={(event) => {
+          const target = event.target as HTMLElement;
+
+          // Make sure we're not clicking on an interactive element
+          if (target.closest("button, a")) {
+            event.preventDefault();
+            return;
+          }
+
+          const input = inputRef.current;
+          if (!input) return;
+
+          requestAnimationFrame(() => {
+            input.focus();
+          });
+        }}
```

Comment:
This handles 2 things. 
- When there is a trailing action button that is clicked, the field doesn't pick up that focus state so the field remains in it's normal state
- When there is an icon in the front without an action, when the icon is clicked the field is focused for the user.